File: gaia-sdk-javascript/src/mqtt/MqttSensorQueue.ts
import * as mqtt from 'mqtt';
import {IQueueOptions} from './QueueOptions';
import {IQueueHeader} from './QueueHeader';
import {ConversationQueueType} from './ConversationQueueType';
import {UserProperties} from 'mqtt-packet';

export class MqttSensorQueue {

    private readonly options: IQueueOptions;
    private client: mqtt.Client;
    private readonly subscriptions: Map<string, QueueCallback> = new Map();

    constructor(options: IQueueOptions) {
        this.options = options;
        this.client = this.connect();
    }

    public connect = () => mqtt.connect(this.options.host, this.options);

    public on(event: string, callback: any) {
        this.client.on(event, callback);
    }

    public end(force: boolean, callback: any) {
        this.client.end(force, callback);
    }

    public publish(conversationQueueType: ConversationQueueType, header: IQueueHeader, payload: object, attributes: object, type: string) {
        const topic = this.getTopic(conversationQueueType, header);
        console.debug(`Sending message to topic ${topic}`);
        const payloadStr = JSON.stringify({attributes, type, payload: payload instanceof Array ? payload[0] : payload});
        const userProperties = Object.assign(header, {
            deviceInstanceId: this.options.deviceInstanceId,
            deviceId: this.options.deviceId,
            userId: this.options.userId
        });
        const opts: IMqttPublishOpts = {properties: {userProperties: JSON.parse(JSON.stringify(userProperties))}};
        this.client.publish(topic, payloadStr, opts, this.mqttCallback(JSON.parse(payloadStr)));
    }

    public subscribe(type: ConversationQueueType, header: IQueueHeader, callback?: QueueCallback) {
        const topic = this.getTopic(type, header);
        if (topic in this.subscriptions.keys()) {
            console.warn(`${topic} already subscribed`);
            return;
        }
        if (callback) {
            console.debug('subscribe to', topic);
            this.client.subscribe(topic, callback);
            this.subscriptions.set(topic, callback);
        }
    }

    public unsubscribe(topic: string) {
        this.client.unsubscribe(topic);
    }

    public subscribeConvContext = (header: IQueueHeader, callback?: QueueCallback) => this.subscribe(ConversationQueueType.CONTEXT, header, callback);
    public subscribeConvNotification = (header: IQueueHeader, callback?: QueueCallback) => this.subscribe(ConversationQueueType.NOTIFICATION, header, callback);
    public subscribeConvInteraction = (header: IQueueHeader, callback?: QueueCallback) => this.subscribe(ConversationQueueType.INTERACTION, header, callback);
    public subscribeConvLogging = (header: IQueueHeader, callback?: QueueCallback) => this.subscribe(ConversationQueueType.LOGGING, header, callback);


    public publishConvInteraction(header: IQueueHeader, interaction: ConvInteraction) {
        if (this.getTopic(ConversationQueueType.INTERACTION, header) in this.subscriptions.keys()) {
            this.subscribeConvInteraction(header, () => {
            });
        }
        this.publish(ConversationQueueType.INTERACTION, header, interaction.payload, interaction.attributes, interaction.type);
    }

    private mqttCallback(payload: object): mqtt.PacketCallback {
        return (error?: Error, packet?: mqtt.Packet) => {
            error ?
                console.error('Failed to publish message', error, packet) :
                console.debug('Successfully published message', payload);
        };
    }

    public callback = (topic: string, payload: object) => {
        if (this.subscriptions.get(topic)) {
            const callback = this.subscriptions.get(topic);
            callback!(payload);
            return;
        }
        console.warn('no callback for topic ' + topic, payload);
    }

    public getTopic = (type: ConversationQueueType, header: IQueueHeader) =>
        `gaia/conversation/${this.options.deviceInstanceId}/${header.channelId}/${type.toLowerCase()}`

}

export declare type QueueCallback = (message: any) => void;

export interface ConvInteraction {
    payload: object;
    attributes: object;
    type: string;
}

interface IMqttPublishOpts extends mqtt.IClientPublishOptions {
    properties?: {
        payloadFormatIndicator?: boolean,
        messageExpiryInterval?: number,
        topicAlias?: number,
        responseTopic?: string,
        correlationData?: Buffer,
        userProperties?: UserProperties,
        subscriptionIdentifier?: number,
        contentType?: string
    };
}

File: gaia-sdk-java/gaia-sdk-mqtt/src/main/kotlin/gaia/sdk/mqtt/MqttSensorQueue.kt
/*
 * Copyright (c) 2016-2020, Leftshift One
 * __________________
 * [2020] Leftshift One
 * All Rights Reserved.
 * NOTICE:  All information contained herein is, and remains
 * the property of Leftshift One and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Leftshift One
 * and its suppliers and may be covered by Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Leftshift One.
 */

package gaia.sdk.mqtt

import com.fasterxml.jackson.databind.ObjectMapper
import com.hivemq.client.mqtt.MqttGlobalPublishFilter.SUBSCRIBED
import com.hivemq.client.mqtt.MqttWebSocketConfig
import com.hivemq.client.mqtt.lifecycle.MqttClientAutoReconnect
import com.hivemq.client.mqtt.mqtt5.Mqtt5Client
import com.hivemq.client.mqtt.mqtt5.Mqtt5RxClient
import com.hivemq.client.mqtt.mqtt5.message.connect.Mqtt5Connect
import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish
import gaia.sdk.api.ISensorQueue
import gaia.sdk.mqtt.async.SdkThreadFactory
import gaia.sdk.mqtt.queue.*
import gaia.sdk.mqtt.queue.ConversationQueueType.*
import gaia.sdk.spi.QueueOptions
import io.reactivex.Completable
import io.reactivex.Flowable
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors.newFixedThreadPool
import java.util.concurrent.TimeUnit

class MqttSensorQueue(private val options: QueueOptions) : ISensorQueue {
    companion object {
        const val GAIA_ROOT_TOPIC = "gaia"
    }

    private val callbacks = ConcurrentHashMap<String, (QueuePayload<ByteArray>) -> Unit>()
    private val logger = LoggerFactory.getLogger(MqttSensorQueue::class.java)
    private val objectMapper = ObjectMapper()

    private val client = initClient()

    fun connect(): Completable {
        return client.connect(Mqtt5Connect.builder().cleanStart(false)
                .noSessionExpiry().keepAlive(10).build())
                .doOnSuccess { handle() }
                .ignoreElement()
    }

    private fun handle() {
        client.publishes(SUBSCRIBED)
                .onErrorResumeNext(Flowable.empty())
                .map { QueueMessage(it) }
                .subscribeOn(options.subscriptionScheduler)
                .subscribe({
                    if (callbacks.containsKey(it.getTopic())) {
                        val contentType = it.getContentType().orElse("application/json")
                        val payload = QueuePayload(contentType, it.getPayloadAsBytes())
                        callbacks[it.getTopic()]!!(payload)
                    }
                }) { it.printStackTrace() }
    }

    fun subscribe(type: IQueueType, header: QueueHeader, consumer: (QueuePayload<ByteArray>) -> Unit): Completable {
        return client.subscribeWith()
                .topicFilter(getTopic(type, header)).noLocal(true)
                .applySubscribe()
                .timeout(options.subscribeTimeout, TimeUnit.SECONDS)
                .doOnSuccess {
                    logger.info("Successfully subscribed")
                    callbacks[getTopic(type, header)] = consumer
                }
                .doOnError { logger.error("Unable to subscribe", it.cause) }
                .ignoreElement()
    }

    fun unsubscribe(type: IQueueType, header: QueueHeader): Completable {
        return client.unsubscribeWith()
                .topicFilter(getTopic(type, header))
                .applyUnsubscribe()
                .timeout(options.subscribeTimeout, TimeUnit.SECONDS)
                .doOnSuccess {
                    logger.info("Successfully unsubscribed")
                    callbacks.remove(getTopic(type, header))
                }
                .doOnError { logger.error("Unable to unsubscribe", it.cause) }
                .ignoreElement()
    }

    fun publish(type: IQueueType, header: QueueHeader, payload: QueuePayload<ByteArray>): Completable {
        val publish = Mqtt5Publish.builder()
                .topic(getTopic(type, header))
                .userProperties()
                .add("identityId", header.identityId.toString())
                .add("userId", UUID.randomUUID().toString())
                .add("deviceId", options.deviceId)
                .add("deviceInstanceId", options.deviceInstanceId)
                .add("channelId", header.channelId?.toString() ?: UUID.randomUUID().toString())
                .applyUserProperties()
                .contentType(payload.contentType)
                .payload(payload.content)
                .build()
        return client.publish(Flowable.just(publish)).ignoreElements()
    }

    private fun initClient(): Mqtt5RxClient {
        var clientBuilder = Mqtt5Client.builder()
                .identifier(options.deviceInstanceId)
                .serverHost(options.host)
                .serverPort(options.port)
                .automaticReconnect(MqttClientAutoReconnect.builder().build())

        if (!options.username.isNullOrEmpty() && !options.password.isNullOrEmpty()) {
            clientBuilder = clientBuilder.simpleAuth()
                    .username(options.username!!)
                    .password(options.password!!.toByteArray())
                    .applySimpleAuth()
        }

        if (options.isSsl) {
            clientBuilder = clientBuilder
                    .sslWithDefaultConfig()
        }

        if (options.isWebsocket) {
            clientBuilder = clientBuilder
                    .webSocketConfig(MqttWebSocketConfig.builder().serverPath(options.websocketPath).build())
        }

        return clientBuilder
                .executorConfig()
                .nettyExecutor(newFixedThreadPool(options.clientThreads, SdkThreadFactory.INSTANCE))
                .applicationScheduler(options.clientScheduler)
                .nettyThreads(options.clientThreads)
                .applyExecutorConfig()
                .addConnectedListener { logger.info("Connected to {}", it.clientConfig.serverAddress) }
                .addDisconnectedListener { logger.error("Error while connecting to mqtt broker", it.cause) }
                .buildRx()
    }

    private fun getTopic(type: IQueueType, header: QueueHeader): String {
        if (type is ConversationQueueType) {
            val deviceInstanceId = options.deviceInstanceId
            val channelId = header.channelId
            return when (type) {
                NOTIFICATION -> "$GAIA_ROOT_TOPIC/conversation/$deviceInstanceId/$channelId/notification"
                INTERACTION -> "$GAIA_ROOT_TOPIC/conversation/$deviceInstanceId/$channelId/interaction"
                CONTEXT -> "$GAIA_ROOT_TOPIC/conversation/$deviceInstanceId/$channelId/context"
                LOGGING -> "$GAIA_ROOT_TOPIC/conversation/$deviceInstanceId/$channelId/logging"
            }
        }
        throw IllegalArgumentException("cannot handle queue type $type")
    }

    fun subscribeConvContext(header: QueueHeader, consumer: (QueuePayload<ConvContext>) -> Unit): Completable {
        return subscribe(CONTEXT, header) {
            consumer(QueuePayload(it.contentType, objectMapper.readValue(it.content, ConvContext::class.java)))
        }
    }

    fun subscribeConvLogging(header: QueueHeader, consumer: (QueuePayload<ConvLogging>) -> Unit): Completable {
        return subscribe(CONTEXT, header) {
            consumer(QueuePayload(it.contentType, objectMapper.readValue(it.content, ConvLogging::class.java)))
        }
    }

    fun subscribeConvNotification(header: QueueHeader, consumer: (QueuePayload<ConvNotification>) -> Unit): Completable {
        return subscribe(CONTEXT, header) {
            consumer(QueuePayload(it.contentType, objectMapper.readValue(it.content, ConvNotification::class.java)))
        }
    }

    fun subscribeConvInteraction(header: QueueHeader, consumer: (QueuePayload<ConvInteraction>) -> Unit): Completable {
        return subscribe(CONTEXT, header) {
            consumer(QueuePayload(it.contentType, objectMapper.readValue(it.content, ConvInteraction::class.java)))
        }
    }

    fun publishConvInteraction(header: QueueHeader, payload: ConvInteraction): Completable {
        val content = QueuePayload("application/json", objectMapper.writeValueAsBytes(payload.content))
        return publish(INTERACTION, header, content)
    }

}



